clash-prelude/src/Clash/Explicit/BlockRam/Model.hs

```diff
@@ -0,0 +1,311 @@
+{-|
+Copyright  :  (C) 2023, QBayLogic B.V.
+License    :  BSD2 (see the file LICENSE)
+Maintainer :  QBayLogic B.V. <devops@qbaylogic.com>
+
+Configurable model for "Clash.Explicit.BlockRam"
+-}
+
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Clash.Explicit.BlockRam.Model where
+
+import Clash.Promoted.Nat (SNat(..), natToNum)
+import Clash.Signal.Bundle (Bundle(bundle))
+import Clash.Signal.Internal
+  (KnownDomain(..), Clock (..), Signal (..), ClockAB (..), clockTicks)
+import Clash.Sized.Index (Index)
+import Clash.XException (NFDataX(..), seqX)
+import Clash.XException.MaybeX (MaybeX(..), toMaybeX, andX)
+
+import Control.Applicative (liftA2)
+import Data.Maybe (fromMaybe)
+import Data.Sequence (Seq)
+import GHC.Stack (HasCallStack)
+import GHC.TypeNats (KnownNat)
+
+import qualified Data.Sequence as Seq
+import qualified Clash.XException.MaybeX as MaybeX
+
+-- | Helper used in 'getConflict'
+data Conflict = Conflict
+  { cfRWA :: !(MaybeX Bool) -- ^ Read/Write conflict for output A
+  , cfRWB :: !(MaybeX Bool) -- ^ Read/Write conflict for output B
+  , cfWW  :: !(MaybeX Bool) -- ^ Write/Write conflict
+  } deriving (Show)
+
+-- | Determines whether there was a write/write or read/write conflict. A conflict
+-- occurs when two ports tried to (potentially, in case of undefined values)
+-- access the same address and one or both tried to write to it. See documentation
+-- of 'Conflict' for more information.
+getConflict ::
+  -- | Port A: enable, write enable, address
+  (MaybeX Bool, MaybeX Bool, MaybeX Int) ->
+  -- | Port B: enable, write enable, address
+  (MaybeX Bool, MaybeX Bool, MaybeX Int) ->
+  -- | 'Just' if there is a (potential) write conflict, otherwise 'Nothing'
+  Maybe Conflict
+getConflict (enA, wenA, addrA) (enB, wenB, addrB)
+  | IsDefined False <- sameAddrX = Nothing
+  | otherwise                    = Just conflict
+ where
+  sameAddrX = liftA2 (==) addrA addrB
+
+  conflict = Conflict
+    { cfRWA = andX enA (andX enB wenB)
+    , cfRWB = andX enB (andX enA wenA)
+    , cfWW  = andX (andX enA enB) (andX wenA wenB)
+    }
+
+-- | Step through a cycle of a TDP Block RAM where only one clock is active. Like
+-- 'writeRam', it accounts for 'Clash.XException.XException' in all values
+-- supplied by the user of the block RAM.
+cycleOne ::
```

Comment:
You're doing quite a lot with `errorX` and `deepErrorX`, and `trueDualPortBlockRam#` has `HasCallStack`, but all of these functions do not. Please sprinkle some more `HasCallStack` so we actually have, well you know, call stacks.

Reply:
Sprinkled!